File: backend/services/bist/fund_allocation.py
"""
TEFAS's fifty-four portfolio lines, collapsed into something a bar can say.

TEFAS publishes a fund's split across 54 instrument codes. Drawn literally that
is 54 segments, most of them a hair wide, and a reader learns nothing from it.
Drawn as a dozen buckets it answers the question the bar exists for: is this an
equity fund, a bond fund, or a money-market fund wearing a different name.

**The rule that decides every bucket:** group by economic exposure where the
field names it, and by wrapper where the wrapper is all the field names. A gold
participation account tracks the gold price, not a bank's deposit rate, so it is
a precious metal and not a deposit — a katılım gold fund holds almost nothing
else, and filing it under deposits would draw it as a money-market fund. A real
estate investment fund's units are real estate. But a plain `yyf` participation
unit could hold anything at all, so "fund units" *is* the honest label.

Two judgement calls that could reasonably go the other way, recorded so they are
argued with rather than rediscovered:

* `eut` (Eurobonds) sits under public debt. Most of the Turkish eurobond float
  is sovereign; corporate issues exist and land in the wrong bucket. Giving them
  their own bucket would spend a thirteenth colour on a rounding error.
* `ymk` (Yabancı Menkul Kıymet) sits under foreign debt. The code is generic,
  but `yhs` already carries foreign equity, so `ymk` is what is left.

Pure: no I/O, no cache, no clock. `fund_service` owns all three.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from services.bist.tefas_client import ALLOCATION_FIELDS

# Where an unmapped field goes. TEFAS added codes in the 2026 rewrite and will
# again; dropping an unknown one would quietly shrink the bar and leave the
# reported total unexplained.
UNKNOWN_BUCKET = "diger"

# Bar order, and deliberately also the adjacency plan: because the order is
# fixed, the pairs that can ever touch are fixed too. Gold and amber are the
# only two fills close in hue and they sit six positions apart, so no bar can
# put them side by side. See `frontend/lib/fund-allocation.ts` for the colours.
BUCKET_ORDER: tuple[str, ...] = (
    "hisse",
    "yabanci_hisse",
    "gayrimenkul",
    "fon",
    "kiymetli_maden",
    "ozel_borclanma",
    "kamu_borclanma",
    "yabanci_borclanma",
    "mevduat",
    "para_piyasasi",
    "turev",
    UNKNOWN_BUCKET,
)

BUCKET_LABELS: dict[str, str] = {
    "hisse": "Hisse senedi",
    "yabanci_hisse": "Yabancı hisse senedi",
    "gayrimenkul": "Gayrimenkul ve girişim",
    "fon": "Fon katılma payları",
    "kiymetli_maden": "Kıymetli madenler",
    "ozel_borclanma": "Özel sektör borçlanma",
    "kamu_borclanma": "Kamu borçlanma",
    "yabanci_borclanma": "Yabancı borçlanma",
    "mevduat": "Mevduat ve katılma hesabı",
    "para_piyasasi": "Repo ve para piyasası",
    "turev": "Türev ve teminat",
    UNKNOWN_BUCKET: "Diğer",
}

_BUCKET_FIELDS: dict[str, tuple[str, ...]] = {
    "hisse": ("hs",),
    "yabanci_hisse": ("yhs",),
    # Unit trusts included: a GYF or GSYF unit is real estate or venture by
    # definition, where a plain fund unit is not.
    "gayrimenkul": ("gas", "gyy", "gsyy", "gykb", "gsykb"),
    "fon": ("yyf", "byf", "ybyf", "fkb"),
    # `khau`/`vmau` are gold accounts and belong to the metal, not to the bank
    # holding it — see the module docstring.
    "kiymetli_maden": ("km", "kmbyf", "kmkba", "kmkks", "khau", "vmau"),
    "ozel_borclanma": ("ost", "fb", "bb", "vdm", "osdb", "osks", "oksyd"),
    "kamu_borclanma": (
        "dt",
        "hb",
        "kibd",
        "kba",
        "kkstl",
        "kksd",
        "kksyd",
        "kks",
        "eut",
        "db",
        "dot",
    ),
    "yabanci_borclanma": ("ybkb", "ybosb", "yba", "ymk"),
    "mevduat": ("vmtl", "vmd", "vm", "khtl", "khd", "kh"),
    "para_piyasasi": ("bpp", "tpp", "r", "tr", "btaa", "btas"),
    "turev": ("t", "vint"),
    UNKNOWN_BUCKET: ("d",),
}

FIELD_BUCKETS: dict[str, str] = {
    field: bucket for bucket, fields in _BUCKET_FIELDS.items() for field in fields
}


@dataclass(frozen=True)
class AllocationLine:
    """One TEFAS instrument line inside a bucket."""

    code: str
    label: str
    weight: float


@dataclass(frozen=True)
class AllocationBucket:
    key: str
    label: str
    weight: float
    lines: tuple[AllocationLine, ...]
    """Largest first. One line means the bucket *is* that line."""


@dataclass(frozen=True)
class AllocationBreakdown:
    """One fund's split, grouped and ready to draw."""

    day: date
    total: float
    """
    Everything TEFAS reported, summed. **Never normalised to 1.**

    Rows land near 1 but rarely on it. Scaling the segments up to fill the bar
    would invent the missing sliver; leaving the remainder as bare track and
    printing the real total lets the page say what it actually knows.
    """
    buckets: tuple[AllocationBucket, ...]
    """In `BUCKET_ORDER`, not by size — a fixed order is what lets two funds'
    bars be read against each other, which is the whole point of the column."""


def group_allocation(weights: dict[str, float], day: date) -> AllocationBreakdown | None:
    """
    Collapse one fund's reported lines into buckets.

    Returns None for a fund with nothing reported. An empty breakdown would draw
    a zero-length bar, and a zero-length bar reads as "this fund holds nothing" —
    a claim, where the truth is that TEFAS published no claim at all.
    """
    if not weights:
        return None

    grouped: dict[str, list[AllocationLine]] = {}
    for code, weight in weights.items():
        if weight <= 0:
            continue
        bucket = FIELD_BUCKETS.get(code, UNKNOWN_BUCKET)
        label = ALLOCATION_FIELDS.get(code, code)
        grouped.setdefault(bucket, []).append(AllocationLine(code=code, label=label, weight=weight))

    if not grouped:
        return None

    buckets: list[AllocationBucket] = []
    for key in BUCKET_ORDER:
        lines = grouped.get(key)
        if not lines:
            continue
        lines.sort(key=lambda line: line.weight, reverse=True)
        buckets.append(
            AllocationBucket(
                key=key,
                label=BUCKET_LABELS[key],
                weight=sum(line.weight for line in lines),
                lines=tuple(lines),
            )
        )

    return AllocationBreakdown(
        day=day,
        total=sum(bucket.weight for bucket in buckets),
        buckets=tuple(buckets),
    )


def bucket_weights(breakdown: AllocationBreakdown) -> dict[str, float]:
    """
    The screener row's form: bucket key to weight, absent meaning not held.

    Sparse and unlabelled on purpose. The board carries this for every fund on
    it, and repeating a dozen Turkish labels thirteen hundred times to say the
    same thing the vocabulary in the response meta already says is most of the
    payload.
    """
    return {bucket.key: bucket.weight for bucket in breakdown.buckets}


def bucket_vocabulary() -> list[dict[str, str]]:
    """The key-to-label table, sent once on the board rather than per row."""
    return [{"key": key, "label": BUCKET_LABELS[key]} for key in BUCKET_ORDER]

File: frontend/lib/fund-allocation.ts
/**
 * Colour and order for the fund allocation bar.
 *
 * The grouping itself is the backend's — `services/bist/fund_allocation.py`
 * decides that a gold participation account is gold rather than a deposit.
 * What is decided here is only what a reader sees: which colour a bucket wears
 * and where in the bar it sits.
 *
 * **The colour is fixed per bucket, not resolved per bar.** This is the one
 * thing that separates this module from `components/ownership/format.ts`, whose
 * `allocationColors` picks colours against the other segments in the same bar.
 * That is right for holdings, which are arbitrary and unbounded per entity. It
 * would be wrong here: with a per-bar rotation, one fund's second segment and
 * another fund's second segment take the same colour while meaning different
 * things, and reading two rows of the screener against each other — the entire
 * reason the column exists — stops working.
 *
 * The order is fixed for the same reason, and it doubles as the adjacency plan:
 * because the sequence cannot change, the pairs that can ever touch cannot
 * either. Gold (`--data-gold`) and amber (`--chart-3`) are the only two fills
 * close in hue, and they sit six positions apart.
 *
 * The one pairing worth designing around is `hisse` beside `fon`. Those are the
 * two commonest lines TEFAS reports — 975 and 1,197 funds of 2,029 — so they
 * end up adjacent on most bars on the board, whatever the order says. `fon`
 * therefore takes terracotta rather than the violet that reads as a shade of
 * the equity indigo beside it, and violet goes to the rare property bucket.
 */

import type { FundAllocationWeights } from '@/lib/bist-api';
import { formatPercent } from '@/lib/bist-format';

/**
 * Bar order. Mirrors `BUCKET_ORDER` in `services/bist/fund_allocation.py`; the
 * server already sends buckets in it, and this is what keeps a row honest if a
 * response ever arrives in another order.
 */
export const FUND_BUCKET_ORDER = [
  'hisse',
  'yabanci_hisse',
  'gayrimenkul',
  'fon',
  'kiymetli_maden',
  'ozel_borclanma',
  'kamu_borclanma',
  'yabanci_borclanma',
  'mevduat',
  'para_piyasasi',
  'turev',
  'diger',
] as const;

/** The bucket an unmapped key falls back to, matching the backend's. */
export const UNKNOWN_BUCKET = 'diger';

/**
 * Hue names the asset family, lightness the sub-kind — so a 96px bar reads as
 * about six families at a glance and the detail legend disambiguates.
 *
 * The dim siblings mix toward `--surface-2` rather than toward `transparent`.
 * The bar's own track is `--surface-2`, so the two look identical today, but
 * only by accident: a transparent segment would shift the moment the bar sat
 * over anything else, and it would stop working entirely in the light theme.
 *
 * Nothing here is green, red, or `--warn`. A green segment on a fund's bar
 * would be read as performance. Precious metals take the app-wide gold token
 * for the same reason bitcoin is bitcoin orange everywhere: the hue names the
 * instrument, so a gold fund is recognisable across the whole board.
 */
export const FUND_BUCKET_COLOR: Record<string, string> = {
  hisse: 'var(--chart-1)',
  yabanci_hisse: 'color-mix(in srgb, var(--chart-1) 55%, var(--surface-2))',
  gayrimenkul: 'var(--chart-7)',
  fon: 'var(--chart-4)',
  kiymetli_maden: 'var(--data-gold)',
  ozel_borclanma: 'var(--chart-2)',
  kamu_borclanma: 'var(--chart-5)',
  yabanci_borclanma: 'color-mix(in srgb, var(--chart-5) 45%, var(--surface-2))',
  mevduat: 'var(--chart-6)',
  para_piyasasi: 'color-mix(in srgb, var(--chart-6) 55%, var(--surface-2))',
  turev: 'var(--chart-3)',
  diger: 'var(--fg-subtle)',
};

export interface FundAllocationSegment {
  key: string;
  label: string;
  /** Fraction of the portfolio. */
  weight: number;
  color: string;
}

export function bucketColor(key: string): string {
  return FUND_BUCKET_COLOR[key] ?? FUND_BUCKET_COLOR[UNKNOWN_BUCKET];
}

/**
 * The bar's segments, in bar order.
 *
 * `labels` is the vocabulary the board sends once in its meta block. A bucket
 * the server added and this file has not heard of keeps that server label and
 * takes the "Diğer" colour rather than being dropped — a bar that silently
 * shrinks when the backend gains a bucket would be worse than an odd colour.
 */
export function allocationSegments(
  allocation: FundAllocationWeights | null | undefined,
  labels: Record<string, string> = {}
): FundAllocationSegment[] {
  if (!allocation) return [];

  const known = FUND_BUCKET_ORDER.filter((key) => key in allocation);
  const unknown = Object.keys(allocation).filter(
    (key) => !(FUND_BUCKET_ORDER as readonly string[]).includes(key)
  );

  return [...known, ...unknown]
    .map((key) => ({
      key,
      label: labels[key] ?? key,
      weight: allocation[key],
      color: bucketColor(key),
    }))
    .filter((segment) => Number.isFinite(segment.weight) && segment.weight > 0);
}

/**
 * What TEFAS reported, summed. Never clamped to 1 — the shortfall is a fact
 * about the filing and the bar leaves it as bare track.
 */
export function allocationTotal(allocation: FundAllocationWeights | null | undefined): number {
  if (!allocation) return 0;
  return Object.values(allocation).reduce(
    (sum, weight) => (Number.isFinite(weight) && weight > 0 ? sum + weight : sum),
    0
  );
}

/**
 * The scalar the screener column sorts on: how much of the fund is equity.
 *
 * A stacked bar has no natural ordering, and "how much of this is stocks" is
 * the risk axis a screener reader is already on — it sorts the board from money
 * market to pure equity in one click. Null only when nothing was reported: a
 * fund that reported and holds no equity is a 0, and the two must not tie.
 */
export function equityWeight(allocation: FundAllocationWeights | null | undefined): number | null {
  if (!allocation) return null;
  return (allocation.hisse ?? 0) + (allocation.yabanci_hisse ?? 0);
}

/** The largest bucket, for a one-glance label. Null when nothing was reported. */
export function dominantBucket(segments: FundAllocationSegment[]): FundAllocationSegment | null {
  if (segments.length === 0) return null;
  // Ties resolve to the earlier segment, which is bar order — stable, and the
  // same answer for two funds holding the same halves.
  return segments.reduce((best, segment) => (segment.weight > best.weight ? segment : best));
}

/**
 * A bucket's weight, printed.
 *
 * The floor is the point. The bar already refuses to let a 0.03% line vanish,
 * and printing it as "%0,0" beside the segment would undo that in words: a
 * reader would take it for a holding of nothing, which is exactly the claim the
 * 2px minimum exists to avoid making.
 */
export function formatWeight(weight: number): string {
  if (weight > 0 && weight < 0.0005) return '<%0,1';
  return formatPercent(weight);
}

/**
 * The bar's tooltip: "Hisse senedi %53,2 · Fon katılma payları %32,0 · …".
 *
 * A stacked bar is unreadable to a screen reader and imprecise to a mouse, so
 * the same sentence serves as both the `title` and the `aria-label`.
 */
export function allocationSummary(segments: FundAllocationSegment[]): string {
  return segments.map((segment) => `${segment.label} ${formatWeight(segment.weight)}`).join(' · ');
}
